lms/djangoapps/courseware/management/commands/migrate_student_module_history.py

```diff
@@ -12,56 +13,82 @@
     """
-    Command to migrate all data from StudentModuleHistoryArchive into StudentModuleHistory.
+    Command to migrate data from StudentModuleHistoryArchive into StudentModuleHistory.
+    Works from largest ID to smallest.
     """
     help = dedent(__doc__).strip()
+    option_list = BaseCommand.option_list + (
+        make_option('-M', '--max-id', type='int', default=None, dest='max_id',
+            help='maximum ID to copy. Defaults to max ID in CSMH.')
+        make_option('-m', '--min_id', type='int', default=0, dest='min_id',
+            help='minimum ID to copy. Defaults to 0.')
+        make_option('-w', '--window', type='int', default=1000, dest='window',
+            help='how many rows to migrate per query. Defaults to 1000.')
+    )
+
+    def handle(self, *arguments, **options):
+        opts, args = _parse_args(arguments, options)
+
+        if options['max_id'] is None:
+            try:
+                max_id = StudentModuleHistory.objects.all().order_by('id')[0].id
+            except IndexError:
+                self.stdout.write("No entries found in StudentModuleHistory, aborting migration.\n")
+                return
+        else:
+            max_id = options['max_id']
+
+        self.migrate_range(options['min_id'], max_id, options['window'])
 
-    @transaction.commit_manually
-    def handle(self, *args, **options):
-        try:
-            max_id = StudentModuleHistory.objects.all().order_by('id')[0].id
-        except IndexError:
-            self.stdout.write("No entries found in StudentModuleHistory, aborting migration.\n")
-            return
 
+    @transaction.commit_manually
+    def _migrate_range(self, min_id, max_id, window):
+        self.stdout.write("Migrating StudentModuleHistoryArchive entries {}-{}\n".format(max_id, min_id))
         start_time = time.time()
 
-        self.stdout.write("Migrating StudentModuleHistoryArchive entries before {}\n".format(max_id))
         archive_entries = (
             StudentModuleHistoryArchive.objects
             .select_related('student_module__student')
             .order_by('-id')
         )
 
-        entry = None
+
+        real_max_id = None
         count = 0
-        window = 100
+        current_max_id = max_id
 
         try:
-            while max_id > 0:
-                entries = archive_entries.filter(id__lt=max_id, id__gte=max_id - window)
+            #what about the entries min_id < x < window when (max_id - min_id) % window != 0?
+            while current_max_id > min_id:
+                #this could overlap into the range below
+                entries = archive_entries.filter(id__lt=current_max_id, id__gte=current_max_id - window)
+
                 new_entries = [StudentModuleHistory.from_archive(entry) for entry in entries]
 
                 StudentModuleHistory.objects.bulk_create(new_entries)
                 count += len(entries)
 
-                if entries:
+                if entries:     #when would this be false?
```

Comment:
When the chunk of data doesn't actually have any history entries in it (in particular, we jump a big gap during the database migration, so that we have space to fill in while the old code is still running on servers)